File: v1/lovart_reverse/execution.py
"""Execution metadata helpers for CLI and MCP command results."""

from __future__ import annotations

from typing import Any

LOCAL = "local"
PREFLIGHT = "preflight"
SUBMIT = "submit"


def annotate(
    data: dict[str, Any],
    execution_class: str,
    *,
    network_required: bool,
    remote_write: bool,
    submitted: bool | None = None,
    cache_used: bool | None = None,
) -> dict[str, Any]:
    result = dict(data)
    result["execution_class"] = execution_class
    result["network_required"] = network_required
    result["remote_write"] = remote_write
    if submitted is not None:
        result["submitted"] = submitted
    if cache_used is not None:
        result["cache_used"] = cache_used
    return result

File: cli/execution.go
package cli

import "github.com/aporicho/lovart/internal/envelope"

const (
	executionLocal     = "local"
	executionPreflight = "preflight"
	executionSubmit    = "submit"
)

func okLocal(data any, cacheUsed ...bool) envelope.Envelope {
	meta := envelope.ExecutionMetadata{
		ExecutionClass:  executionLocal,
		NetworkRequired: false,
		RemoteWrite:     false,
	}
	if len(cacheUsed) > 0 {
		meta.CacheUsed = boolPtr(cacheUsed[0])
	}
	return envelope.OKWithMetadata(data, meta)
}

func okPreflight(data any, cacheUsed ...bool) envelope.Envelope {
	meta := envelope.ExecutionMetadata{
		ExecutionClass:  executionPreflight,
		NetworkRequired: true,
		RemoteWrite:     false,
	}
	if len(cacheUsed) > 0 {
		meta.CacheUsed = boolPtr(cacheUsed[0])
	}
	return envelope.OKWithMetadata(data, meta)
}

func okSubmit(data any, submitted bool) envelope.Envelope {
	return envelope.OKWithMetadata(data, envelope.ExecutionMetadata{
		ExecutionClass:  executionSubmit,
		NetworkRequired: true,
		RemoteWrite:     submitted,
		Submitted:       boolPtr(submitted),
	})
}

func okRemoteWrite(data any) envelope.Envelope {
	return envelope.OKWithMetadata(data, envelope.ExecutionMetadata{
		ExecutionClass:  executionSubmit,
		NetworkRequired: true,
		RemoteWrite:     true,
	})
}

func okPreflightSubmission(data any, submitted bool) envelope.Envelope {
	return envelope.OKWithMetadata(data, envelope.ExecutionMetadata{
		ExecutionClass:  executionPreflight,
		NetworkRequired: true,
		RemoteWrite:     false,
		Submitted:       boolPtr(submitted),
	})
}

func boolPtr(value bool) *bool {
	return &value
}
